combined_robot_hw_tests/CMakeLists.txt

```diff
@@ -43,18 +62,26 @@ if(CATKIN_ENABLE_TESTING)
     test/cm_test.cpp
   )
   target_link_libraries(combined_robot_hw_cm_test ${PROJECT_NAME} ${catkin_LIBRARIES})
-
 endif()
 
-# Install
+
+#############
+## Install ##
+#############
+
+# Install headers
 install(DIRECTORY include/${PROJECT_NAME}/
-  DESTINATION ${CATKIN_PACKAGE_INCLUDE_DESTINATION})
+  DESTINATION ${CATKIN_PACKAGE_INCLUDE_DESTINATION}
+)
 
-# Install library
-install(TARGETS ${PROJECT_NAME}
+# Install targets
+install(TARGETS ${PROJECT_NAME} combined_robot_hw_dummy_app
```

Comment:
the dummy app is only for tests, dont think we need to install

Reply:
Got it. In that case, should I also remove the install for `${PROJECT_NAME}`? It seems to me we should either install all targets or none, and none seems to make more sense. Should I also remove the exported headers and targets from the `catkin_package()` call?

Same for headers and plugins - Do we need to install anything at all?

These test-only packages still feel quite odd to me. Why are they separate packages at all, and more importantly, why are they released and dependencies of `ros_control`? There are only a couple of packages with 'test' in their name in [rosdistro](https://github.com/ros/rosdistro/blob/master/melodic/distribution.yaml), and as far as I can tell we have almost the only packages that are just rostest & rosunit tests.

Reply:
I guess it depends on how one views it, should these be installable to then run as standalone tests or not? I think that since most of the framework has strong API dependencies internally, these would need to be recompiled anyways and if needed, they can be locally compiled. 

The original reasoning for these test-only packages was to have an all-round integration suite, these packages export dummy robothw components, controllers, etc. 

I'd opt for the choice of not installing or exporting headers from these packages.

I'm also in favor of removing them from the ros_control metapackage however I am not entirely sure we can get rid of them in the rosdistro file, the release scripts will probably want to update them all and bloom will look for them too. Let's try gutting them first.

Reply:
> The original reasoning for these test-only packages was to have an all-round integration suite, these packages export dummy robothw components, controllers, etc.

Do you mean so that people who write plugins and extensions have some tools to test their code? That would be cool, I saw a couple packages like that when I looked through rosdistro. I agree that they're not really offering that as they are though.

> I'm also in favor of removing them from the ros_control metapackage however I am not entirely sure we can get rid of them in the rosdistro file

Agreed. Un-releasing packages doesn't seem possible or desirable, even if they're mostly useless to the outside world. Should we remove them from the metapackage in Melodic (And maybe Kinetic), or only in Noetic+ to stay consistent with the rosdistro file? I would lean towards removing it today.

> I'd opt for the choice of not installing or exporting headers from these packages

Sounds good. I will also remove the plugin export.

---

A final note: Shall I open an issue to track this so we can keep improving the situation? Whether that means updating the metapackage or pulling the tests into their corresponding packages (leaving empty packages for now, to be removed in Noetic) or whatever else we decide.

Fixing the CMakeLists.txt is certainly in scope for this PR, but playing with it any further feels like it deserves its own PR.

Reply:
Looking at `test_roscpp` and the other `test_XXX` packages in `ros_comm` as an example for these "test-packages," it seems we are on the right track by gutting the CMakeLists.txt of any installs and exports, and eventually halting distribution of the packages altogether.

Following the example from those packages, we should also be using `EXCLUDE_FROM_ALL` on our `add_executable` and `add_library` targets. We should also move all of our `include_directory` and `add_executable`/`add_library` stuff into the `if(CATKIN_ENABLE_TESTING)`, leaving just the stuff like `catkin_package` outside the if.

I will follow the example of these packages and update our `_tests` packages accordingly.

Reply:
Fixed in 6f4648de

Reply:
Looks like tests fail if we don't install plugins. Will investigate.

Reply:
I'm able to reproduce the issue reliably using `industrial_ci` locally, but I haven't been able to reproduce it using `catkin run_tests` directly, even doing my best to reproduce the steps `industrial_ci` takes.

Regardless, I suspect the issue may be related to `catkin_tools` [incorrectly sourcing the devel space](https://github.com/catkin/catkin_tools/issues/397) when running tests. I haven't fully investigated, but it seems the easiest solution is probably to just install these plugins. Unless you've got a better idea, I'm going to revert back to installing just the plugins.

Reply:
Re-installed in 2864284b

Reply:
Test still fails because now the plugin can't find its corresponding library, since it's not installed.

It seems we're stuck installing all the bits related to plugins, since tests are run out of the install space rather than the devel space.

Reply:
Update: I can reproduce the issue if I run the tests out of the install space.

This behaviour is consistent with `industrial_ci`. Essentially:

```
catkin config --install
catkin build
catkin run_tests
```

As I understand the issue, since `industrial_ci` uses the install space and `catkin_tools` doesn't switch to the devel space when running tests (as discussed in the issue linked above), the only solution is to install the plugins and corresponding libraries required by the tests.

Reply:
Please open a follow-up issue for this and let's continue the discussion there.

Reply:
Will do.

Reply:
Issues #409 and #410 opened.